pkg/sql/exec/distinct.go

```diff
@@ -0,0 +1,154 @@
+// Copyright 2018 The Cockroach Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+// implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+package exec
+
+// sortedDistinctInt64Op runs a distinct on the column in sortedDistinctCol,
+// writing the result to the bool column in outputColIdx by or'ing the
+// difference between the current and last value in the column with what's
+// already in the output column. this has the effect of setting the output
+// column to true when the input is distinct.
+type sortedDistinctInt64Op struct {
+	input Operator
+
+	// sortedDistinctCol is the index of the column to distinct upon.
+	sortedDistinctCol int
+
+	// outputColIdx is the index of the boolean output column in the input batch.
+	outputColIdx int
+
+	// Set to true at runtime when we've seen the first row. Distinct always
+	// outputs the first row that it sees.
+	foundFirstRow bool
+
+	// lastVal is the last value seen by the operator, so that the distincting
+	// still works across batch boundaries.
+	lastVal int64 // template
+}
+
+var _ Operator = &sortedDistinctInt64Op{}
+
+func (p *sortedDistinctInt64Op) Init() {}
+
+func (p *sortedDistinctInt64Op) Next() ColBatch {
+	batch := p.input.Next()
+	if batch.Length() == 0 {
+		return batch
+	}
+	outputCol := batch.ColVec(p.outputColIdx).Bool()
+	col := batch.ColVec(p.sortedDistinctCol).Int64()
+
+	// We always output the first row.
+	lastVal := p.lastVal
+	sel := batch.Selection()
+	if !p.foundFirstRow {
+		if sel != nil {
+			lastVal = col[sel[0]]
+			outputCol[sel[0]] = true
+		} else {
+			lastVal = col[0]
+			outputCol[0] = true
+		}
+	}
+
+	startIdx := uint16(0)
+	if !p.foundFirstRow {
+		startIdx = 1
+	}
+
+	n := batch.Length()
+	if sel != nil {
+		// Bounds check elimination.
+		sel = sel[startIdx:n]
+		for s := 0; s < len(sel); s++ {
+			i := sel[s]
+			unique := col[i] != lastVal
+			outputCol[i] = outputCol[i] || unique
+			lastVal = col[i]
+		}
+	} else {
+		// Bounds check elimination.
+		col = col[startIdx:n]
+		outputCol = outputCol[startIdx:n]
+		for i := 0; i < len(col); i++ {
+			unique := col[i] != lastVal
```

Comment:
[nit] I would add a comment here saying that it's important that `unique` is defined separately..
Or maybe rewrite as
```go
v := col[i]
outputCol[i] = outputCol[i] || (v != lastVal)
lastVal = v
```